File: ruby/grade-school/grade_school.rb
class School
  def initialize
    @grades = Hash.new { |h, g| h[g] = [] }
  end

  def grade(g)
    @grades[g].sort
  end

  def roster
    @grades.to_a.sort_by(&:first).map(&:last).flat_map(&:sort)
  end

  def add(name, grade)
    return false if @grades.any? { |_, names| names.include? name }
    @grades[grade] << name
    true
  end
end

File: javascript/grade-school/grade-school.js
export class GradeSchool {
    constructor() {
        this._roster = {};
    }

    roster() {
        const roster = {};
        for (const grade in this._roster) {
            roster[grade] = Array.from(this._roster[grade]);
        }
        return roster;
    }

    add(student, grade) {
        if (!(grade in this._roster)) {
            this._roster[grade] = [student];
        } else {
            this._roster[grade].push(student);
            this._roster[grade].sort();
        }
    }

    grade(grade) {
        return grade in this._roster ? Array.from(this._roster[grade]) : [];
    }
}
